### PR TITLE
Clarify dynamic tool lazy-load timing

### PR DESCRIPTION
**Pull request by basnijholt**
## Summary

- Replace dynamic-tool lazy-loading wording that told agents to wait for the next request.
- Continue the same AI response after `load_tool()` / `unload_tool()` manager calls so updated tool state is available in a later tool-call step.
- Keep the continuation limit visible even when the final empty model response contains a visible tool trace.
- Preserve model-prompt metadata on continuation attempts without re-appending the original user request.
- Update docs, generated skill references, and regression coverage.

## Testing

- `uv sync --all-extras`
- `uv run pre-commit run --all-files`
- `uv run pytest tests/test_dynamic_tool_continuation.py tests/test_dynamic_toolkits.py tests/test_ai_user_id.py::TestUserIdPassthrough::test_ai_response_continues_after_dynamic_tool_load tests/test_ai_user_id.py::TestUserIdPassthrough::test_stream_agent_response_continues_after_dynamic_tool_load tests/test_ai_user_id.py::TestUserIdPassthrough::test_ai_response_returns_message_after_dynamic_tool_limit -n 0 --no-cov -q` (`43 passed`)
- `uv run pytest -n auto --no-cov -q` (`8541 passed, 30 skipped, 56 warnings`)

### CODE REVIEW

File: tests/test_dynamic_toolkits.py [hunk 1]
```diff
@@ -557,7 +557,11 @@ def test_dynamic_tools_manager_loads_unloads_searches_and_respects_sticky_initia
...
 
-    assert _tool_payload(manager.load_tool("sleep"))["status"] == "loaded"
+    loaded_payload = _tool_payload(manager.load_tool("sleep"))
+    assert loaded_payload["status"] == "loaded"
+    assert loaded_payload["takes_effect"] == "later_tool_call_step"
+    assert "same response" in loaded_payload["message"]
+    assert "next request" not in loaded_payload["message"]
     assert _tool_payload(manager.load_tool("sleep"))["status"] == "already_loaded"
     assert _tool_payload(manager.unload_tool("shell"))["status"] == "sticky"
     assert _tool_payload(manager.unload_tool("sleep"))["status"] == "unloaded"
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply tests/test_dynamic_toolkits.py]: Fixed in 9df62b120: added field-level assertions for the already_loaded and unloaded dynamic-tool payloads, including takes_effect and the removal of next-request wording.

File: src/mindroom/custom_tools/dynamic_tools.py [hunk 2]
```diff
@@ -148,8 +148,11 @@ def _load_tool_response(self, tool_name: str, result: LoadToolResult) -> str:
                 "already_loaded",
                 tool_name=tool_name,
                 loaded_tools=loaded_tools,
-                takes_effect="next_request",
-                message=f"Tool '{tool_name}' is already loaded for this session.",
+                takes_effect="later_tool_call_step",
+                message=(
+                    f"Tool '{tool_name}' is already loaded. Continue the same response and call it "
+                    "in a later tool-call step."
+                ),
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/custom_tools/dynamic_tools.py]: Fixed in 9df62b120: changed the already_loaded message to use “Continue the same task” for consistency with the rest of the dynamic-tool guidance.

File: src/mindroom/custom_tools/dynamic_tools.py [hunk 3]
```diff
@@ -255,8 +261,11 @@ def unload_tool(self, tool_name: str) -> str:
             "unloaded",
             tool_name=tool_name,
             loaded_tools=saved_loaded_tools,
-            takes_effect="next_request",
-            message=f"Tool '{tool_name}' will be removed on the next request in this session.",
+            takes_effect="later_tool_call_step",
+            message=(
+                f"Tool '{tool_name}' is unloaded. Continue the same response without calling it "
+                "in later tool-call steps."
+            ),
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/custom_tools/dynamic_tools.py]: Fixed in 9df62b120: changed the unloaded message to use “Continue the same task” for consistency with the rest of the dynamic-tool guidance.

**Comment by basnijholt** [comment]
Addressed the remaining Greptile note from https://github.com/mindroom-ai/mindroom/pull/1289#issuecomment-4704909715 in 78c592a5f. The continuation notice is now status-neutral for no-op/error dynamic-tool calls, and both streaming and non-streaming paths now emit/record a visible fallback message if repeated dynamic-tool calls hit the continuation limit without producing user-visible content. Added regression coverage for both paths.

**Comment by basnijholt** [comment]
Reworked the dynamic-tool continuation implementation in 85be5c2d8. The behavior is the same, but the core ai.py change is now a narrow post-run continuation hook instead of the previous large nested rewrite: PR-facing ai.py diff is now 166 additions / 31 deletions, with the decision/prompt/limit logic moved to src/mindroom/dynamic_tool_continuation.py.

Also fixed the cancellation edge case I noticed during the rewrite: if a continuation run is cancelled, interrupted history now preserves the earlier dynamic-tool trace without duplicating it. Added a regression test for that path.

Verification:
- uv run pytest tests/test_dynamic_tool_continuation.py tests/test_dynamic_toolkits.py tests/test_ai_user_id.py -n 0 --no-cov -q -> 189 passed
- touched-file ruff check / ruff format --check / ty check / git diff --check -> passed
- uv run pre-commit run --all-files still fails only on unrelated untracked src/mindroom/team_execution_attempt.py / tests/test_team_execution_attempt.py issues already present in the worktree; module privacy now passes.

**Comment by basnijholt** [comment]
Follow-up pushed in e7c5991de after comparing Claude Code's lazy tool loading semantics.

Changes:
- Tightened dynamic-tool wording to say the updated schema is available after the tool result is processed, and that the loaded tool should be called in a later tool-call step, not the same parallel batch.
- Changed hidden continuation prompts so successful load/already-loaded statuses mention the updated tool schema, while unknown/unavailable/non-loaded states tell the model to continue with the current dynamic tool state and explain or choose another available deferred tool instead of assuming the requested tool is usable.

Verification:
- uv run pytest tests/test_dynamic_tool_continuation.py tests/test_dynamic_toolkits.py tests/test_ai_user_id.py -n 0 --no-cov -q -> 190 passed
- touched-file ruff check / ruff format --check / ty check / git diff --check -> passed
- uv run pre-commit run --all-files remains blocked only by the unrelated untracked src/mindroom/team_execution_attempt.py / tests/test_team_execution_attempt.py issues.

**Comment by basnijholt** [comment]
Addressed the Greptile finding from https://github.com/mindroom-ai/mindroom/pull/1289#issuecomment-4704909715 in b364178b7.

Confirmed this was a real bug: if a streaming dynamic-tool continuation was cancelled, the nested continuation recorded the correct interrupted state first, then the outer stream cancellation handler could overwrite it with only the first-run state.

Fix: the outer streaming call now detects cancellation propagated from a nested continuation and re-raises without running its own interrupted-state recorder/persist path, preserving the nested continuation's partial text and tool list.

Added regression test: tests/test_ai_user_id.py::TestUserIdPassthrough::test_stream_agent_response_continuation_task_cancel_preserves_second_run_tools

Verification:
- uv run pytest tests/test_ai_user_id.py::TestUserIdPassthrough::test_stream_agent_response_continuation_task_cancel_preserves_second_run_tools -x -n 0 --no-cov -q
- uv run pytest tests/test_dynamic_tool_continuation.py tests/test_dynamic_toolkits.py tests/test_ai_user_id.py -n 0 --no-cov -q
- uv run ruff check src/mindroom/ai.py tests/test_ai_user_id.py
- uv run ruff format --check src/mindroom/ai.py tests/test_ai_user_id.py
- uv run ty check src/mindroom/ai.py tests/test_ai_user_id.py
- git diff --check

Full uv run pre-commit run --all-files is still blocked by unrelated untracked src/mindroom/team_execution_attempt.py and tests/test_team_execution_attempt.py work: ty reports the missing fallback_prompt argument there, and vulture reports unused methods there. The hooks relevant to this PR's touched files passed.

File: src/mindroom/ai.py [hunk 4]
```diff
@@ -1057,236 +1062,285 @@ async def ai_response(  # noqa: C901, PLR0912, PLR0915
...
+                        compaction_lifecycle=compaction_lifecycle,
+                        delegation_depth=delegation_depth,
+                        refresh_scheduler=refresh_scheduler,
+                        system_enrichment_items=system_enrichment_items,
                         include_openai_compat_guidance=include_openai_compat_guidance,
-                    ),
-                    pipeline_timing=pipeline_timing,
+                        timing_scope=timing_scope,
+                        model_prompt=model_prompt,
+                        **_current_sender_id_kwargs(
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/ai.py]: Fixed in 782f5fa18: continuation attempts now pass only the model-prompt tail after the original raw prompt, so model metadata is preserved without re-appending the user request. Added non-streaming and streaming regression assertions.
